### PR TITLE
[Proposal] Add parent_child_checkbox as part of Flutter Plugins or Widget

### LINKED ISSUE #90107: [Proposal] Add parent_child_checkbox as part of Flutter Plugins or Widget
## Description
Parent-Child Checkbox is a type of checkbox where we can establish a hierarchy in Checkboxes.
You can create a Parent Checkbox which has multiple checkboxes for their children! The value of the parent checkbox depends on the value of children checkboxes and vice-versa.<br>
If we clicked on Parent Checkbox, all children checkboxes will be selected too. 

## Reason to add the plugin:
- Parent-Child-Checkbox is a common functionality needed by developers
- Android have it by default in their [SDK](https://developer.android.com/reference/kotlin/androidx/compose/material/package-summary#TriStateCheckbox(androidx.compose.ui.state.ToggleableState,kotlin.Function0,androidx.compose.ui.Modifier,kotlin.Boolean,androidx.compose.foundation.interaction.MutableInteractionSource,androidx.compose.material.CheckboxColors))
- Integral part of [Material Design](https://material.io/develop/android/components/checkboxes)
- Part of most of the frameworks

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### PR DESCRIPTION
**Pull request by AbhishekDoshi26**
Parent-Child Checkbox is a type of checkbox where we can establish a hierarchy in Checkboxes.
You can create a Parent Checkbox which has multiple checkboxes for their children! The value of the parent checkbox depends on the value of children checkboxes and vice-versa.

If we clicked on Parent Checkbox, all children checkboxes will be selected too.

Issues fixed: #90107 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: examples/parent_child_checkbox/lib/main.dart [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import 'dart:developer';
```
Comment (goderbauer) [inline examples/parent_child_checkbox/lib/main.dart:1]: There should be some sample code in the API docs for the widget that shows how to use it (look for `{@tool dartpad` sections in our existing code base for examples). With that, this dedicated example project in the example directory can be removed.

File: packages/flutter/lib/src/widgets/parent_child_checkbox.dart [hunk 2]
```diff
@@ -0,0 +1,183 @@
+import 'package:flutter/material.dart';
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/parent_child_checkbox.dart:1]: missing license header

File: packages/flutter/lib/src/widgets/parent_child_checkbox.dart [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+///Main widget of Parent Child Checkbox.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/parent_child_checkbox.dart:3]: space after `///`. (here and elsewhere)

The first line should be a one sentence description of what the widget is and not just repeat its name.

Also, the doc needs to be expanded with more details about what this widget does and how to use it. Ideally with some sample code.

File: packages/flutter/lib/src/widgets/parent_child_checkbox.dart [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+///Parent Child Checkbox is a type of checkbox where we can establish hierarchy in Checkboxes.
+class ParentChildCheckbox extends StatefulWidget {
+  ///Text Widget to specify the Parent checkbox
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/parent_child_checkbox.dart:6]: docs should be full sentences and end with a `.`

Also, most docs in this file need improvements as they are not very informative (see also https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-useless-documentation).

File: packages/flutter/lib/src/widgets/parent_child_checkbox.dart [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    required this.parent,
+    required this.children,
+    this.parentCheckboxColor,
+    this.childrenCheckboxColor,
+  });
+
+  /// Map which shows whether particular parent is selected or not.
+  ///
+  /// Example: {'Parent 1' : true, 'Parent 2' : false} where
+  /// Parent 1 and Parent 2 will be 2 separate parents if you are using multiple ParentChildCheckbox in your code.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/parent_child_checkbox.dart:33]: nit: keep docs to about 80chars/line

File: packages/flutter/lib/src/widgets/parent_child_checkbox.dart [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+  /// Default value will be false for all specified parents
+  static Map<String?, bool?> _isParentSelected = {};
+
+  /// Getter to get whether particular parent is selected or not.
+  ///
+  /// Example: {'Parent 1' : true, 'Parent 2' : false} where
+  /// Parent 1 and Parent 2 will be 2 separate parents if you are using multiple ParentChildCheckbox in your code.
+  ///
+  /// Default value will be false for all specified parents
+  static get isParentSelected => _isParentSelected;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/parent_child_checkbox.dart:44]: The static API on this class is odd.

File: packages/flutter/lib/src/widgets/parent_child_checkbox.dart [hunk 7]
```diff
@@ -0,0 +1,183 @@
+import 'package:flutter/material.dart';
+
+///Main widget of Parent Child Checkbox.
+///Parent Child Checkbox is a type of checkbox where we can establish hierarchy in Checkboxes.
+class ParentChildCheckbox extends StatefulWidget {
+  ///Text Widget to specify the Parent checkbox
+  final Text? parent;
+
+  ///List<Text> Widgets to specify the Children checkboxes
+  final List<Text>? children;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/parent_child_checkbox.dart:10]: Forcing these to be Text widgets is against Flutter style: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-interleaving-multiple-concepts-together

File: packages/flutter/lib/src/widgets/parent_child_checkbox.dart (hunk 6 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/parent_child_checkbox.dart:44]: Similarly, the use of Strings in the API is error-prone.

**Comment by goderbauer** [comment]
This PR is also missing tests and the checks are not passing.

Overall, such a widget could also easily be provided in a separate package on pub.dev and doesn't have to live in the framework.

**Comment by AbhishekDoshi26** [comment]
> This PR is also missing tests and the checks are not passing.
> 
> Overall, such a widget could also easily be provided in a separate package on pub.dev and doesn't have to live in the framework.

I agree on this and I have already published a package for the same. As this was a material design component, I created PR for the same

**Pull request closed without merging**